builder/googlecompute/tunnel_driver.go

```diff
@@ -0,0 +1,80 @@
+// +build !windows
+
+package googlecompute
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"os/exec"
+	"strings"
+	"syscall"
+	"time"
+)
+
+func NewTunnelDriver() TunnelDriver {
+	return &TunnelDriverLinux{}
+}
+
+type TunnelDriverLinux struct {
+	cmd *exec.Cmd
+}
+
+func (t *TunnelDriverLinux) StartTunnel(cancelCtx context.Context, tempScriptFileName string) error {
+	// set stdout and stderr so we can read what's going on.
+	var stdout, stderr bytes.Buffer
+
+	cmd := exec.CommandContext(cancelCtx, tempScriptFileName)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	err := cmd.Start()
+	log.Printf("Waiting 30s for tunnel to create...")
+	if err != nil {
+		err := fmt.Errorf("Error calling gcloud sdk to launch IAP tunnel: %s",
+			err)
+		return err
+	}
+	// Wait for tunnel to launch and gather response. TODO: do this without
+	// a sleep.
+	time.Sleep(30 * time.Second)
+
+	// Track stdout.
+	sout := stdout.String()
+	if sout != "" {
+		log.Printf("[start-iap-tunnel] stdout is:")
+	}
+
+	log.Printf("[start-iap-tunnel] stderr is:")
+	serr := stderr.String()
+	log.Println(serr)
+	if strings.Contains(serr, "ERROR") {
+		errIdx := strings.Index(serr, "ERROR:")
+		return fmt.Errorf("ERROR: %s", serr[errIdx+7:])
+	}
+	// Store successful command on step so we can access it to cancel it
+	// later.
+	t.cmd = cmd
+	return nil
+}
+
+func (t *TunnelDriverLinux) StopTunnel() {
+	if t.cmd != nil && t.cmd.Process != nil {
+		log.Printf("Cleaning up the IAP tunnel...")
+		// Why not just cmd.Process.Kill()?  I'm glad you asked. The gcloud
+		// call spawns a python subprocess that listens on the port, and you
+		// need to use the process _group_ id to kill this process and its
+		// daemon child. We create the group ID with the syscall.SysProcAttr
+		// call inside the retry loop above, and then store that ID on the
+		// command so we can destroy it here.
+		err := syscall.Kill(-t.cmd.Process.Pid, syscall.SIGKILL)
```

Comment:
if I call cmd.Process.Kill or kill it using the ctx, the child process which is actually listening on the port doesn't die. I have to kill the whole process group. Do you know of a more graceful way to do this? I could try sending a SIGINT to the process group instead. 🤔

Reply:
So I was reading the golang code and it's super weird because when you call process.kill or cancel the context it calls for:

```go
// https://github.com/golang/go/blob/82f29898ea60e7d92001bf1e4c0d48bc1c8aed0d/src/os/exec_posix.go#L65-L67
func (p *Process) kill() error {
	return p.Signal(Kill)
}
```

Which in turn calls:
```go
// https://github.com/golang/go/blob/82f29898ea60e7d92001bf1e4c0d48bc1c8aed0d/src/os/exec_unix.go#L71
    syscall.Kill(p.Pid, s)
```

Reply:
Ah right, yes I think SIGINTing the parent process would be better yes ! I would _think_ it is the right solution because it sort of simulates the `ctrl+c` a terminal user would do to terminate the proxy, and the parent process should cleanup after itself.

Reply:
Sigint does work, but I still need to send to the whole process group not just the parent process.